horovod/common/operations.cc

```diff
@@ -1755,5 +1755,21 @@ Status EnqueueJoin(std::shared_ptr<OpContext> context,
   return status;
 }
 
+// Contexts and controller must be initialized and the background thread
+// must be running before this function is called.
+Status CallBarrier(int32_t process_set_id) {
```

Comment:
Is it safe to call this function directly from the PyTorch thread? I fear that this will be problematic at least when single-threaded MPI is used as the controller. Horovod's overall design is laid out such that a single persistent thread handles all the collective communication between Horovod processes.

Then I would expect that `process_set.controller->Barrier(Communicator::GLOBAL);` should be called in the Horovod background thread. Then there would need to be an `EnqueueBarrier()` function etc. as we have for Join and the more complex ops.

This comment in operations.cc explains this a bit:
https://github.com/horovod/horovod/blob/ddd36d5be1ac8071dab74daa2883dca679b65bc8/horovod/common/operations.cc#L358

Reply:
This is my concern as well. @Tixxx @chongxiaoc can you elaborate on why this isn't handled in the Horovod background thread?

Reply:
Thanks for pointing this out. I have been mainly using openmpi which has been shipped with thread_multiple by default for a while. looks like we still need to support single threaded cases. It makes sense to move this call to the main thread that calls mpi_init then.

Reply:
@maxhgerlach @tgaddair  I added a new barrier request to do barrier in the background thread.